File: 3 - LongestSubstrWithoutRepeatingChars/Solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if (s.length <=1){
        return s.length;
    }
    let charSet = {};
    maxLength = 0;
    let left = 0;
    for(let right=0; right < s.length; right++){
        let c = s.charAt(right);
        if (c in charSet){
            left = Math.max(left, charSet[c] + 1);
        }
        charSet[c] = right;
        maxLength = Math.max(maxLength, right - left + 1);
    }
    return maxLength;
};

File: 3 - LongestSubstrWithoutRepeatingChars/Solution.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.length()<=1){
            return s.length();
        }
        int maxLength = 0;
        int left = 0;
        unordered_map<char, int> charMap;
        for (int right = 0; right < s.length(); right++){
            char c = s[right];
            if (charMap.find(c) != charMap.end()){
                left = max(left, charMap[c]+1);
            }
            charMap[c] = right;
            maxLength = max(maxLength, right - left + 1);
        }
        return maxLength;
    }
};
